expression/expression.go

```diff
@@ -217,17 +217,43 @@ func (sf *ScalarFunction) ToString() string {
 	return result
 }
 
-// NewFunction creates a new scalar function.
-func NewFunction(funcName string, retType *types.FieldType, args ...Expression) (*ScalarFunction, error) {
+// NewFunction creates a new scalar function or constant.
+func NewFunction(funcName string, retType *types.FieldType, args ...Expression) (Expression, error) {
 	f, ok := evaluator.Funcs[funcName]
 	if !ok {
 		log.Errorf("Function %s is not implemented.", funcName)
 		return nil, nil
 	}
+
 	if len(args) < f.MinArgs || (f.MaxArgs != -1 && len(args) > f.MaxArgs) {
 		return nil, evaluator.ErrInvalidOperation.Gen("number of function arguments must in [%d, %d].",
 			f.MinArgs, f.MaxArgs)
 	}
+
+	canConstantFolding := true
+	datums := make([]types.Datum, 0, len(args))
+
+	for _, expr := range args {
```

Comment:
rand() has no args, but it should not be folded.
mysql> select *, rand() from t1;
+------+------+--------------------+
| c1   | c2   | rand()             |
+------+------+--------------------+
|    2 |    3 | 0.8671364708497529 |
|    4 |    4 | 0.1637872002681598 |
+------+------+--------------------+